warning: comparing floating point with == or != is unsafe [-Wfloat-equal]

We like to build network-manager-applet's libnm with many compiler warnings enabled. Now that Gtk 4.0 drags in graphene, we now get some extra ones:

```
[lkundrak@belphegor linux]$ cat test.c 
#include <gtk/gtk.h>
[lkundrak@belphegor linux]$ cc $(pkg-config --cflags gtk+-4.0) -Wfloat-equal -c test.c
In file included from /usr/include/graphene-1.0/graphene.h:35,
                 from /usr/include/gtk-4.0/gsk/gsktypes.h:25,
                 from /usr/include/gtk-4.0/gsk/gskrenderer.h:26,
                 from /usr/include/gtk-4.0/gsk/gsk.h:24,
                 from /usr/include/gtk-4.0/gtk/gtkwidget.h:33,
                 from /usr/include/gtk-4.0/gtk/gtkapplication.h:27,
                 from /usr/include/gtk-4.0/gtk/gtkwindow.h:33,
                 from /usr/include/gtk-4.0/gtk/gtkdialog.h:32,
                 from /usr/include/gtk-4.0/gtk/gtkaboutdialog.h:30,
                 from /usr/include/gtk-4.0/gtk/gtk.h:31,
                 from test.c:1:
/usr/include/graphene-1.0/graphene-simd4f.h: In function ‘graphene_simd4f_is_zero3’:
/usr/include/graphene-1.0/graphene-simd4f.h:1730:36: warning: comparing floating point with == or != is unsafe [-Wfloat-equal]
 1730 |   return graphene_simd4f_get_x (v) == 0.f &&
      |                                    ^~
/usr/include/graphene-1.0/graphene-simd4f.h:1731:36: warning: comparing floating point with == or != is unsafe [-Wfloat-equal]
 1731 |          graphene_simd4f_get_y (v) == 0.f &&
      |                                    ^~
/usr/include/graphene-1.0/graphene-simd4f.h:1732:36: warning: comparing floating point with == or != is unsafe [-Wfloat-equal]
 1732 |          graphene_simd4f_get_z (v) == 0.f;
      |                                    ^~
/usr/include/graphene-1.0/graphene-simd4f.h: In function ‘graphene_simd4f_is_zero2’:
/usr/include/graphene-1.0/graphene-simd4f.h:1749:36: warning: comparing floating point with == or != is unsafe [-Wfloat-equal]
 1749 |   return graphene_simd4f_get_x (v) == 0.f &&
      |                                    ^~
/usr/include/graphene-1.0/graphene-simd4f.h:1750:36: warning: comparing floating point with == or != is unsafe [-Wfloat-equal]
 1750 |          graphene_simd4f_get_y (v) == 0.f;
      |                                    ^~
In file included from /usr/include/graphene-1.0/graphene.h:36,
                 from /usr/include/gtk-4.0/gsk/gsktypes.h:25,
                 from /usr/include/gtk-4.0/gsk/gskrenderer.h:26,
                 from /usr/include/gtk-4.0/gsk/gsk.h:24,
                 from /usr/include/gtk-4.0/gtk/gtkwidget.h:33,
                 from /usr/include/gtk-4.0/gtk/gtkapplication.h:27,
                 from /usr/include/gtk-4.0/gtk/gtkwindow.h:33,
                 from /usr/include/gtk-4.0/gtk/gtkdialog.h:32,
                 from /usr/include/gtk-4.0/gtk/gtkaboutdialog.h:30,
                 from /usr/include/gtk-4.0/gtk/gtk.h:31,
                 from test.c:1:
/usr/include/graphene-1.0/graphene-simd4x4f.h: In function ‘graphene_simd4x4f_inverse’:
/usr/include/graphene-1.0/graphene-simd4x4f.h:900:35: warning: comparing floating point with == or != is unsafe [-Wfloat-equal]
  900 |   if (graphene_simd4f_get_x (det) != 0.f)
      |                                   ^~
[lkundrak@belphegor linux]$ 
```
I have no idea how to avoid the warning, or even whether comparing a float to zero is actually a bad idea, but we'd be super thankful if it could be avoided.

Thanks
Lubo

--- a/meson.build
+++ b/meson.build
@@ -96,6 +96,7 @@ else
     '-Wcast-align',
     '-Wno-unused-local-typedefs',
     '-Werror=float-conversion',
+    '-Werror=float-equal',
     '-Werror=redundant-decls',
     '-Werror=missing-prototypes',
     '-Werror=missing-declarations',

--- a/meson.build
+++ b/meson.build
@@ -171,6 +171,15 @@ conf.set('HAVE_SINCOSF',
   description: 'Define if sincosf() is available',
 )
 
+conf.set('HAVE_ISINFF',
+  cc.has_function('isinff',
+    prefix: '#include <math.h>',
+    args: [ '-D_GNU_SOURCE' ],
+    dependencies: mathlib,
+  ),
+  description: 'Define is isinff() is available',
+)
+
 # Debugging
 debug_flags = []
 buildtype = get_option('buildtype')

--- a/src/graphene-matrix.c
+++ b/src/graphene-matrix.c
@@ -1862,7 +1862,7 @@ matrix_decompose_2d (const graphene_matrix_t *m,
 
   angle = atan2f (row0y, row0x);
 
-  if (angle != 0.f)
+  if (fabsf (angle) > FLT_EPSILON)
     {
       double sn = -row0y, cs = row0x;
       double m11 = row0x, m12 = row0y;
@@ -2066,9 +2066,9 @@ graphene_matrix_interpolate (const graphene_matrix_t *a,
         }
 
       /* Do not rotate "the long way around" */
-      if (rotate_a == 0.f)
+      if (fabs (rotate_a) <= DBL_EPSILON)
         rotate_a = 360;
-      if (rotate_b == 0.f)
+      if (fabs (rotate_b) <= DBL_EPSILON)
         rotate_b = 360;
 
       if (fabs (rotate_a - rotate_b) > 180)

--- a/src/graphene-simd4f.c
+++ b/src/graphene-simd4f.c
@@ -1176,10 +1176,10 @@ graphene_simd4f_t
 (graphene_simd4f_reciprocal) (graphene_simd4f_t v)
 {
   graphene_simd4f_t s = {
-    v.x != 0.f ? 1.0f / v.x : 0.f,
-    v.y != 0.f ? 1.0f / v.y : 0.f,
-    v.z != 0.f ? 1.0f / v.z : 0.f,
-    v.w != 0.f ? 1.0f / v.w : 0.f
+    fabsf (v.x) > FLT_EPSILON ? 1.0f / v.x : 0.f,
+    fabsf (v.y) > FLT_EPSILON ? 1.0f / v.y : 0.f,
+    fabsf (v.z) > FLT_EPSILON ? 1.0f / v.z : 0.f,
+    fabsf (v.w) > FLT_EPSILON ? 1.0f / v.w : 0.f
   };
   return s;
 }
@@ -1200,10 +1200,10 @@ graphene_simd4f_t
 (graphene_simd4f_rsqrt) (graphene_simd4f_t v)
 {
   graphene_simd4f_t s = {
-    v.x != 0.f ? 1.0f / sqrtf (v.x) : 0.f,
-    v.y != 0.f ? 1.0f / sqrtf (v.y) : 0.f,
-    v.z != 0.f ? 1.0f / sqrtf (v.z) : 0.f,
-    v.w != 0.f ? 1.0f / sqrtf (v.w) : 0.f
+    fabsf (v.x) > FLT_EPSILON ? 1.0f / sqrtf (v.x) : 0.f,
+    fabsf (v.y) > FLT_EPSILON ? 1.0f / sqrtf (v.y) : 0.f,
+    fabsf (v.z) > FLT_EPSILON ? 1.0f / sqrtf (v.z) : 0.f,
+    fabsf (v.w) > FLT_EPSILON ? 1.0f / sqrtf (v.w) : 0.f
   };
   return s;
 }
@@ -1252,10 +1252,10 @@ graphene_simd4f_t
                        const graphene_simd4f_t b)
 {
   graphene_simd4f_t s = {
-    b.x != 0.f ? a.x / b.x : 0.f,
-    b.y != 0.f ? a.y / b.y : 0.f,
-    b.z != 0.f ? a.z / b.z : 0.f,
-    b.w != 0.f ? a.w / b.w : 0.f
+    fabsf (b.x) > FLT_EPSILON ? a.x / b.x : 0.f,
+    fabsf (b.y) > FLT_EPSILON ? a.y / b.y : 0.f,
+    fabsf (b.z) > FLT_EPSILON ? a.z / b.z : 0.f,
+    fabsf (b.w) > FLT_EPSILON ? a.w / b.w : 0.f
   };
   return s;
 }
@@ -1367,24 +1367,52 @@ graphene_simd4f_t
   return graphene_simd4f_init (-s.x, s.y, -s.z, s.w);
 }
 
+static inline bool
+approx_equal (float a,
+              float b,
+              float epsilon)
+{
+#ifdef HAVE_ISINFF
+  if (isinff (a) && isinff (b))
+    return true;
+#else
+  if (fpclassify (a) == FP_INFINITE && fpclassify (b) == FP_INFINITE)
+    return true;
+#endif
+
+  float diff = fabsf (a - b);
+  if (isnan (diff))
+    return true;
+
+  if (diff <= epsilon)
+    return true;
+
+  float abs_a = fabsf (a);
+  float abs_b = fabsf (b);
+
+  float largest = abs_b > abs_a ? abs_b : abs_a;
+
+  return diff <= largest * epsilon;
+}
+
 bool
 (graphene_simd4f_cmp_eq) (const graphene_simd4f_t a,
                           const graphene_simd4f_t b)
 {
-  return a.x == b.x &&
-         a.y == b.y &&
-         a.z == b.z &&
-         a.w == b.w;
+  return approx_equal (a.x, b.x, FLT_EPSILON) &&
+         approx_equal (a.y, b.y, FLT_EPSILON) &&
+         approx_equal (a.z, b.z, FLT_EPSILON) &&
+         approx_equal (a.w, b.w, FLT_EPSILON);
 }
 
 bool
 (graphene_simd4f_cmp_neq) (const graphene_simd4f_t a,
                            const graphene_simd4f_t b)
 {
-  return a.x != b.x ||
-         a.y != b.y ||
-         a.z != b.z ||
-         a.w != b.w;
+  return !approx_equal (a.x, b.x, FLT_EPSILON) ||
+         !approx_equal (a.y, b.y, FLT_EPSILON) ||
+         !approx_equal (a.z, b.z, FLT_EPSILON) ||
+         !approx_equal (a.w, b.w, FLT_EPSILON);
 }
 
 bool

--- a/src/graphene-simd4f.h
+++ b/src/graphene-simd4f.h
@@ -31,6 +31,7 @@
 /* needed for memcpy() */
 #include <string.h>
 #include <math.h>
+#include <float.h>
 
 #include "graphene-config.h"
 #include "graphene-macros.h"
@@ -1727,9 +1728,9 @@ graphene_simd4f_is_zero4 (const graphene_simd4f_t v)
 static inline bool
 graphene_simd4f_is_zero3 (const graphene_simd4f_t v)
 {
-  return graphene_simd4f_get_x (v) == 0.f &&
-         graphene_simd4f_get_y (v) == 0.f &&
-         graphene_simd4f_get_z (v) == 0.f;
+  return fabsf (graphene_simd4f_get_x (v)) <= FLT_EPSILON &&
+         fabsf (graphene_simd4f_get_y (v)) <= FLT_EPSILON &&
+         fabsf (graphene_simd4f_get_z (v)) <= FLT_EPSILON;
 }
 
 /**
@@ -1746,8 +1747,8 @@ graphene_simd4f_is_zero3 (const graphene_simd4f_t v)
 static inline bool
 graphene_simd4f_is_zero2 (const graphene_simd4f_t v)
 {
-  return graphene_simd4f_get_x (v) == 0.f &&
-         graphene_simd4f_get_y (v) == 0.f;
+  return fabsf (graphene_simd4f_get_x (v)) <= FLT_EPSILON &&
+         fabsf (graphene_simd4f_get_y (v)) <= FLT_EPSILON;
 }
 
 /**

--- a/src/graphene-simd4x4f.h
+++ b/src/graphene-simd4x4f.h
@@ -969,7 +969,7 @@ graphene_simd4x4f_inverse (const graphene_simd4x4f_t *m,
   const graphene_simd4f_t d0 = graphene_simd4f_mul (r1_sum, r0);
   const graphene_simd4f_t d1 = graphene_simd4f_add (d0, graphene_simd4f_merge_high (d0, d0));
   const graphene_simd4f_t det = graphene_simd4f_sub (d1, graphene_simd4f_splat_y (d1));
-  if (graphene_simd4f_get_x (det) != 0.f)
+  if (fabsf (graphene_simd4f_get_x (det)) >= FLT_EPSILON)
     {
       const graphene_simd4f_t invdet = graphene_simd4f_splat_x (graphene_simd4f_div (graphene_simd4f_splat (1.0f), det));
 

--- a/src/graphene-triangle.c
+++ b/src/graphene-triangle.c
@@ -343,7 +343,7 @@ graphene_triangle_get_uv (const graphene_triangle_t *t,
   dot12 = graphene_vec3_dot (&v1, &v2);
 
   denom = dot00 * dot11 - dot01 * dot01;
-  if (denom == 0.f)
+  if (fabsf (denom) <= FLT_EPSILON)
     return false;
 
   inv_denom = 1.f / denom;

--- a/src/graphene-vectors.c
+++ b/src/graphene-vectors.c
@@ -325,7 +325,7 @@ void
 graphene_vec2_normalize (const graphene_vec2_t *v,
                          graphene_vec2_t       *res)
 {
-  if (graphene_vec2_length (v) != 0.f)
+  if (fabsf (graphene_vec2_length (v)) > FLT_EPSILON)
     res->value = graphene_simd4f_normalize2 (v->value);
   else
     res->value = graphene_simd4f_init_zero ();
@@ -915,7 +915,7 @@ void
 graphene_vec3_normalize (const graphene_vec3_t *v,
                          graphene_vec3_t       *res)
 {
-  if (graphene_vec3_length (v) != 0.f)
+  if (fabsf (graphene_vec3_length (v)) > FLT_EPSILON)
     res->value = graphene_simd4f_normalize3 (v->value);
   else
     res->value = graphene_simd4f_init_zero ();
@@ -1667,7 +1667,7 @@ void
 graphene_vec4_normalize (const graphene_vec4_t *v,
                          graphene_vec4_t       *res)
 {
-  if (graphene_vec4_length (v) != 0.f)
+  if (fabsf (graphene_vec4_length (v)) > FLT_EPSILON)
     res->value = graphene_simd4f_normalize4 (v->value);
   else
     res->value = graphene_simd4f_init_zero ();

